### ISSUE TITLE
Error when setting exported custom resource field to be a new resource

### ISSUE THREAD

**Issue by es1024** [godotengine/godot#66862]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0.beta2

### System information

Ubuntu 20.04

### Issue description

When attempting to set the value of an exported custom resource field of type `TestResource` (extending `Resource`) by using the "New TestResource" option through the editor, the following error is emitted:
```
Failed to set a resource of the type 'Resource' because this EditorResourcePicker only accepts 'TestResource' and its derivatives.
```

### Steps to reproduce

1. Create a custom resource with class name `TestResource` extending `Resource` with a single exported field.
```gdscript
extends Resource
class_name TestResource
@export var field := ''
```
2. Create a custom node type with class name `TestNode` extending `Node` with a single exported field of type `TestResource`.
```gdscript
extends Node
class_name TestNode
@export var node: TestResource
```
4. Create a node of type `TestNode`.
5. Click on the created node. The `Node` field should be `<empty>`.
6. Try to set the `Node` field using the "New TestResource" option. The observed error will be emitted.

### Minimal reproduction project

[resource-test.zip](https://github.com/godotengine/godot/files/9702669/resource-test.zip)

**Comment by codingvessel** [comment]
Can confirm Same bug for Windows 10, Godot 4.0.beta2

**Comment by michaelgundlach** [comment]
Same in 4.0 beta 4, on MacOS 12.3.1, Apple M1

**Comment by sfreed141** [comment]
I'm seeing the same error, but also with an additional bug where any nested resources are *deleted* when the resource script is modified. I'm on Windows 11, 4.0 beta8.

Details and repro steps:
The original repro project does _not_ repro for me. But, if I add a new resource `test_resource2.gd`:
```
extends Resource
class_name TestResource2

@export var field2: int = 100
```
and modify `test_resource.gd` to export a variable of that new resource type:
```
extends Resource

class_name TestResource

@export var field := ''
@export var res : TestResource2  // <--- added field
```
then I see the `Failed to set a resource of the type 'Resource' because this EditorResourcePicker only accepts 'TestResource2' and its derivatives.` error. Worse though, if you create the nested resource through the editor, save, and then edit the resource script (`test_resource.gd`) then the nested resource is deleted (the resource field will be `<empty>`).

![godot-nested-resource-bug](https://user-images.githubusercontent.com/6880264/208203795-d619dc71-4c98-4108-ba9b-657dd8dc93cc.gif)

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/69356 (possible duplicate?).

**Comment by aMytho** [comment]
I'm also having this issue on 4.0 beta 10.

I have an exported array of dialogue resources. I can make a few dialogue resources via the editor, but if I change the size of the array and add a new resource, the items are reset.

**Comment by muno777** [comment]
Also encountering this with beta 10.

**Comment by PrinceMerluza** [comment]
Still exists on beta11

**Comment by asdfer-1234** [comment]
Also encountering this with beta 12.

**Comment by derkork** [comment]
This is still a thing in v4.0.beta16.mono.official [518b9e580]

**Comment by Shadowblitz16** [comment]
I just ran into this issue which makes me sad because the custom resources were making my life easier
![image](https://user-images.githubusercontent.com/2453707/215293338-7a08e4bb-cc03-44e6-a625-80a28e7d685c.png)

Please make sure you guys fix this with tool scripts too.
I don't know if it's a issue but I kinda need to draw where my engine thrusters are

**Comment by animentork** [comment]
Can confirm it still happens with v4.0.rc1.official [8843d9ad3].

**Comment by rodeyfeld** [comment]
Also getting this error in v4.0.rc.1 official

**Comment by GibionMillak** [comment]
Also getting this error in v4.0.rc.2 official

**Comment by Andrew1431** [comment]
I'm still getting this, it seems like a viable way to do things. I'm on latest 4.1.2. Am I doing something wrong?

![image](https://github.com/godotengine/godot/assets/9736372/0bce249b-3371-45ba-994d-61c4cbaa66f0)

**Comment by Andrew1431** [comment]
Restarting the editor fixed the problem.

**Comment by lucassene** [comment]
I'm not getting the error, but the exported variable is always null for me, even though it has a Resource assigned to it in the scene:

-> Inspector with the Custom Resource assigned to the exported variable:
![image](https://github.com/godotengine/godot/assets/16777494/1fffbaaa-47a6-473e-ad1b-9286760125f4)

-> Remote view of the same Node with the exported variable null:
![image](https://github.com/godotengine/godot/assets/16777494/5a348e12-3192-4440-b123-b709f235d228)

**Comment by AndreiSchapov** [comment]
Why is this closed? Just happened to me in Godot 4.2 on Mac M1

**Comment by peck94** [comment]
This issue still exists as of Godot 4.2.1 on Ubuntu 22.04.

**Comment by mabysh** [comment]
Same for me, v4.2.1.stable.flathub on Ubuntu 23.10

**Comment by shokenz** [comment]
I try solution mentioned above, like:
- Restarting the editor
- Resource _init() must have default values for all parameters.
- Instanciate resource and assign to exported resource field.

 But none of them works.
<br>
While I found out that this problem **only happen when custom resource export PackedScene field.**

If I export path of PackedScene instead of PackedScene itself and load PackedScene with load(path) later. There is no problem at least for my case.
<br>
Example of way with problem:
```
class_name CustomResource extends Resource
@export var scene:PackedScene
```

Example of way works for me:
```
class_name CustomResource extends Resource
@export_file var scene_path:String
var scene:PackedScene:
    get:
        return load(scene_path)
```

**Comment by AndreiSchapov** [comment]
> his is what worked for me. Resource `_init` **must** have default values for all parameters.
> Then instantiate resource on export:

I think this won't work for scenarios when I want the same resource instance shared across multiple scenes. This is what I'm trying to do right now and for some reason exported resource is ok in one scene, while in the other scene it's null

**Comment by Ashenguard** [comment]
And also I did encounter this bug... If I use `@export var test: MyExtendedResource` and set it in the editor when playing the game it is null!

```
Godot v4.2.1.stable - Windows 10.0.18363 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1050 (NVIDIA; 31.0.15.5176) - Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz (12 Threads)
```

**Comment by iRyzor** [comment]
The bug is still present in Godot 4.3.dev6

**Comment by twisteddragons** [comment]
Yeah its still here for me on v4.2.2.stable.official. 

The workaround @mabysh gave worked for me. 

It seems the problem is if you have a _init function with one or more parameters with no default value.

An _init with default values for all its parameters, or an _init with no parameters, or having no _init at all _**should**_ fix the issue.

**Comment by PalomaNeri** [comment]
I had the same error just now. (Godot - v4.2.2 - stable)
I solved it by saving the work and restarting the editor.

**Comment by edvein-rin** [comment]
@YuriSizov sorry for tagging, but this issue is still present and many people encounter it including me.

**Comment by YuriSizov** [comment]
@edvein-rin Hey! I'm not a part of this project anymore, so I can't help you. The issue should be reopened, though, given the number of reports confirming it as not fixed.

**Comment by huesonmcgee** [comment]
I too get this in 4.3stable.  Not really sure how to use custom resources if I can't add them to scenes for use.

**Comment by theloneplant** [comment]
Also seeing this issue on 4.3 stable

**Comment by roman-sereda** [comment]
Having the same bug, I don't have `_init` function in resource, but I have a variable: 
```
var node: Node3D
```
the resource is null when I set it using `@export` from inspector, but if I set some default value for this variable in Resource - it works
```
var node: Node3D = Node3D.new()
```

**Comment by bdrelling** [comment]
> Yeah its still here for me on v4.2.2.stable.official.
> 
> The workaround [@mabysh](https://github.com/mabysh) gave worked for me.
> 
> It seems the problem for me was having a _init function with one or more parameters with no default value.
> 
> An _init with default values for all its parameters, or an _init with no parameters, or having no _init at all _**should**_ fix the issue.

@twisteddragons thanks for sharing this, this is exactly what the problem was for me too.

I'm on 4.3 stable and still seeing this issue. 😢

**Comment by Jestus** [comment]
I just wanted to chime in with my own experience with this issue:

I have a base `Item` extending `Resource`.

I then have a few different trees of items branching off of it:

```
                             Item
                             __|__
BallisticRangedWeapon               LaserRangedWeapon
             _|_                            _|_
BallisticPistol BallisticRifle   LaserPistol   LaserRifle
```
All of the LaserPistols, BallisticPistol (etc) resources I create are fine and are NOT subject to the above `null` issue

HOWEVER, I have the same base class and the following extension tree:

```
                                      Item
                                        |
                                       Ammo
                                      __|__
                        BallisticAmmo        LaserAmmo
```
This tree has caused problems. `LaserAmmo` and `BallisticAmmo` custom created objects are changed to `null` at game start.
I have no idea why this is causing problems and the above tree is not.

Following all of these suggestions did not work. I spent ~3hr yesterday meticulously going through the files and cannot figure out why these are being set to `null`. However, removing the lower level of the tree and doing this, DOES work:
```
                                       Item
                                        |
                                       Ammo
```
So I just added a DamageType field to `Ammo` and made them both `Ammo` rather than extending from it.

This is to say, however: this is quite bad! Why are resources randomly being set to `null` in certain extension scenarios but not others?
It makes me quite concerned - will I make a `BallisticMinigun` or something in the above tree, and suddenly that will break too? It makes me feel uncertain about how reliable extending any of these resources really is.

I would suggest that this fix should be relatively high priority as it appears to greatly impede developer design.

**Comment by ariefield** [comment]
On 4.4.stable I was encountering the issue others have mentioned of a custom exported resource appearing to be set/loaded properly, but then it becomes `null` at runtime.

The workaround of setting the default value to be a new instance did NOT work for me. i.e.
```
@export var output: Item = Item.new()
```
This resulted in all instances of this property having only their default values at runtime, regardless of what I set the exported resource to be.

What DID work for me was setting the default for this property to be a preloaded instance of the resource, i.e.
```
@export var output: Item = preload("res://entities/items/weapons/stone_sword.tres")
```

I encountered some very strange behavior while finding my way to this workaround - it's clear that something is wrong here. 

I'll echo the sentiment that this deserves to be prioritized, given that custom `Resource` definitions are very likely to be the backbone of any larger-scale projects.
